project/src/com/google/daggerquery/server/Server.java

```diff
@@ -0,0 +1,73 @@
+package com.google.daggerquery.server;
+
+import com.google.common.collect.ImmutableList;
+import com.google.daggerquery.executor.QueryExecutor;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.server.RoutingHandler;
+import io.undertow.Undertow;
+import io.undertow.util.HttpString;
+import io.undertow.util.StatusCodes;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Deque;
+import java.util.stream.Collectors;
+
+public class Server {
+
+  private static final int PORT = 4921;
+  private static final String HOST = "localhost";
+
+  /**
+   * Starts a server that accepts a single <b>GET</b> request with the specified query.
+   */
+  public static void main(String[] args) {
+    RoutingHandler routingHandler = new RoutingHandler();
+    routingHandler.add("GET", "/daggerquery/{query}", Server::executeQuery);
+
+    Undertow undertowServer = Undertow.builder()
+        .addHttpListener(PORT, HOST)
+        .setHandler(routingHandler)
+        .build();
+    undertowServer.start();
+  }
+
+  /**
+   * Extracts the query from {@link HttpServerExchange} and tries to execute it.
+   *
+   * <p>If http response didn't contain <b>query</b> parameter then this method does nothing.
+   *
+   * <p>Otherwise it executes query and if it was successful sends correct subgraphs from the
+   * origin binding graphs as the result.
+   *
+   * <p>For <b>deps</b> query returns a list with a string representation of source node's dependencies.
+   * For <b>allpaths</b> and <b>somepath</b> returns a formatted {@link com.google.daggerquery.executor.models.Query.Path}
+   * representation with the node names separated by an arrow.
+   */
+  private static void executeQuery(HttpServerExchange exchange) {
+    exchange.getResponseHeaders().put(new HttpString("Access-Control-Allow-Origin"), "*");
+    exchange.getResponseHeaders().put(new HttpString("Access-Control-Allow-Methods"), "GET");
+
+    Deque<String> deque = exchange.getQueryParameters().get("query");
+    if (deque == null) {
+      return;
+    }
+
+    String[] args = deque.stream().filter(value -> value.length() > 0).toArray(String[]::new);
+    try {
+      ImmutableList<String> results = QueryExecutor.execute(args);
+      ByteBuffer[] buffers = results
+        .stream()
+        .map(result -> ByteBuffer.wrap(result.getBytes()))
+        .toArray(ByteBuffer[]::new);
```

Comment:
I'm not sure if you'd have difficulty decoding this from the client side. If it works then great. If not, maybe you could try packaging this as a JSON array (e.g. [use gson to convert Java to JSON](https://stackoverflow.com/a/36634293), [set content type to application/json](https://www.stubbornjava.com/posts/handling-content-types-with-undertow#application-json))

Reply:
I just checked and everything is decoded correctly, **but** as I can see, it reads the content of the response as one line and the elements are not separated by any character. So we cannot understand where the first binding name ends, it is a problem(

In this case JSON can help. But as I see now [there is no special method](https://undertow.io/javadoc/1.3.x/io/undertow/io/Sender.html) for sending JSONs - or we can pack it as a `ByteBuffer` object again?

Reply:
`new Gson().toJson(...)` returns a `String`, which you can just pass on to `exchange.getResponseSender().send(...)`

Reply:
Exactly, sorry for this clarification. I did it, it looks very simple and much better 🎀